File: benchmark/benchmark.jl
# benchmark.jl

using Revise
using BenchmarkTools
using StaticArrays
using ChenSignatures
using Printf
using Dates
using DelimitedFiles
using YAML

# -------- config loading --------

function load_config()
    config_path = joinpath(@__DIR__, "benchmark_config.yaml")
    @assert isfile(config_path) "Config file not found: $config_path"

    cfg = YAML.load_file(config_path)

    Ns       = get(cfg, "Ns", [150, 1000, 2000])
    Ds       = get(cfg, "Ds", [2, 6, 7, 8])
    Ms       = get(cfg, "Ms", [4, 6])
    path_str = get(cfg, "path_kind", "linear")
    runs_dir = get(cfg, "runs_dir", "runs")
    repeats  = get(cfg, "repeats", 5)
    operations_raw = get(cfg, "operations", ["signature", "logsignature"])

    path_kind = Symbol(path_str)
    operations = Symbol.(operations_raw)

    return (Ns = Ns, Ds = Ds, Ms = Ms,
            path_kind = path_kind, runs_dir = runs_dir,
            repeats = repeats, operations = operations)
end

# -------- path generators --------

# linear: [t, 2t, 2t, ...]
function make_path_linear_svec(d::Int, N::Int)
    ts = range(0.0, stop=1.0, length=N)
    [SVector{d,Float64}(ntuple(i -> (i == 1 ? t : 2t), d)) for t in ts]
end

function make_path_linear_matrix(d::Int, N::Int)
    ts = range(0.0, stop=1.0, length=N)
    path = Matrix{Float64}(undef, N, d)
    path[:, 1] .= ts
    for j in 2:d
        path[:, j] .= 2 .* ts
    end
    return path
end

# sinusoid: [sin(2π·1·t), sin(2π·2·t), ...]
function make_path_sin_svec(d::Int, N::Int)
    ts = range(0.0, stop=1.0, length=N)
    ω = 2π
    [SVector{d,Float64}(ntuple(i -> sin(ω * i * t), d)) for t in ts]
end

function make_path_sin_matrix(d::Int, N::Int)
    ts = range(0.0, stop=1.0, length=N)
    ω = 2π
    path = Matrix{Float64}(undef, N, d)
    for j in 1:d
        path[:, j] .= sin.(ω * j .* ts)
    end
    return path
end

function make_path(d::Int, N::Int, kind::Symbol, path_type::Symbol)
    if kind === :linear
        return path_type === :Matrix ? make_path_linear_matrix(d, N) : make_path_linear_svec(d, N)
    elseif kind === :sin
        return path_type === :Matrix ? make_path_sin_matrix(d, N) : make_path_sin_svec(d, N)
    else
        error("Unknown path_kind: $kind (expected :linear or :sin)")
    end
end

# -------- one benchmark case (Julia only) --------

function bench_case(d::Int, m::Int, N::Int, path_kind::Symbol, op::Symbol, path_type::Symbol, repeats::Int)
    path = make_path(d, N, path_kind, path_type)
    tensor_type = ChenSignatures.Tensor{Float64}

    # Determine path type string for output
    path_type_str = path_type === :Matrix ? "Matrix" : "Vector{SVector}"

    # Helper closures for the two operations
    run_sig() = signature_path(tensor_type, path, m)
    run_logsig() = ChenSignatures.log(signature_path(tensor_type, path, m))

    # Select function and warmup
    if op === :signature
        method_name = "signature_path"
        run_sig()
        # Note: We must interpolate local functions with $ for BenchmarkTools
        t_jl = @belapsed $run_sig() evals=1 samples=repeats
        a_jl = @allocated run_sig()
    elseif op === :logsignature
        method_name = "log"
        run_logsig()
        t_jl = @belapsed $run_logsig() evals=1 samples=repeats
        a_jl = @allocated run_logsig()
    else
        error("Unknown operation: $op")
    end

    t_ms      = t_jl * 1000
    alloc_KiB = a_jl / 1024

    return (N = N,
            d = d,
            m = m,
            path_kind = path_kind,
            operation = op,
            language = "julia",
            library = "ChenSignatures.jl",
            method = method_name,
            path_type = path_type_str,
            t_ms = t_ms,
            alloc_KiB = alloc_KiB)
end

# -------- sweep + write grid to file --------

function run_bench()
    cfg = load_config()
    Ns, Ds, Ms = cfg.Ns, cfg.Ds, cfg.Ms
    path_kind  = cfg.path_kind
    runs_dir   = cfg.runs_dir
    repeats    = cfg.repeats
    operations = cfg.operations

    # Benchmark both path types
    path_types = [:Matrix, :VectorSVector]

    println("Running Julia benchmark with config:")
    println("  path_kind  = $path_kind")
    println("  Ns         = $(Ns)")
    println("  Ds         = $(Ds)")
    println("  Ms         = $(Ms)")
    println("  operations = $(operations)")
    println("  path_types = $(path_types)")
    println("  runs_dir   = \"$runs_dir\"")
    println("  repeats    = $repeats")

    results = NamedTuple[]

    for N in Ns, d in Ds, m in Ms, op in operations, ptype in path_types
        push!(results, bench_case(d, m, N, path_kind, op, ptype, repeats))
    end

    # Base runs dir (for standalone Julia usage)
    base_runs_path = joinpath(@__DIR__, runs_dir)
    isdir(base_runs_path) || mkpath(base_runs_path)

    # Allow orchestrator to override output path
    custom_csv = get(ENV, "BENCHMARK_OUT_CSV", "")
    if !isempty(custom_csv)
        file = custom_csv
        run_path = dirname(file)
        isdir(run_path) || mkpath(run_path)
    else
        ts   = Dates.format(Dates.now(), "yyyymmdd_HHMMSS")
        file = joinpath(base_runs_path, "run_julia_$ts.csv")
    end

    header = ["N", "d", "m", "path_kind", "operation", "language", "library", "method", "path_type", "t_ms", "alloc_KiB"]
    data = Array{Any}(undef, length(results) + 1, length(header))
    data[1, :] = header

    for (i, r) in enumerate(results)
        data[i + 1, 1] = r.N
        data[i + 1, 2] = r.d
        data[i + 1, 3] = r.m
        data[i + 1, 4] = String(r.path_kind)
        data[i + 1, 5] = String(r.operation)
        data[i + 1, 6] = r.language
        data[i + 1, 7] = r.library
        data[i + 1, 8] = r.method
        data[i + 1, 9] = r.path_type
        data[i + 1, 10] = r.t_ms
        data[i + 1, 11] = r.alloc_KiB
    end

    writedlm(file, data, ',')

    println("============================================================")
    println("Benchmark grid written to: $file")
    return file
end

run_bench()

File: benchmark/benchmark.py
# benchmark.py
import os
import csv
import math
import time
import tracemalloc
import ast
import sys
from pathlib import Path
from datetime import datetime

import numpy as np

# Try importing both libraries
try:
    import iisignature
    HAS_IISIG = True
except ImportError:
    HAS_IISIG = False
    print("Warning: iisignature not available", file=sys.stderr)

try:
    import pysiglib
    HAS_PYSIGLIB = True
except ImportError:
    HAS_PYSIGLIB = False
    print("Warning: pysiglib not available", file=sys.stderr)

# -------- simple YAML-ish config loader --------

def load_simple_yaml(path: Path) -> dict:
    cfg = {}
    with path.open("r", encoding="utf-8") as f:
        for line in f:
            line = line.split("#", 1)[0].strip()
            if not line or ":" not in line:
                continue
            key, value = line.split(":", 1)
            key = key.strip()
            value = value.strip()
            if not value:
                continue

            if value.startswith('"') and value.endswith('"'):
                cfg[key] = value[1:-1]
            elif value.startswith("["):
                cfg[key] = ast.literal_eval(value)
            else:
                try:
                    cfg[key] = int(value)
                except ValueError:
                    cfg[key] = value
    return cfg

def load_config():
    script_dir = Path(__file__).resolve().parent
    config_path = script_dir / "benchmark_config.yaml"
    if not config_path.is_file():
        raise FileNotFoundError(f"Config file not found: {config_path}")

    raw = load_simple_yaml(config_path)

    Ns = raw.get("Ns", [150, 1000, 2000])
    Ds = raw.get("Ds", [2, 6, 7, 8])
    Ms = raw.get("Ms", [4, 6])
    path_kind = raw.get("path_kind", "linear")
    runs_dir = raw.get("runs_dir", "runs")
    repeats = int(raw.get("repeats", 5))
    logsig_method = raw.get("logsig_method", "O")
    operations = raw.get("operations", ["signature", "logsignature"])

    path_kind = path_kind.lower()
    if path_kind not in ("linear", "sin"):
        raise ValueError(f"Unknown path_kind '{path_kind}', expected 'linear' or 'sin'.")

    return {
        "Ns": Ns,
        "Ds": Ds,
        "Ms": Ms,
        "path_kind": path_kind,
        "runs_dir": runs_dir,
        "repeats": repeats,
        "logsig_method": logsig_method,
        "operations": operations,
    }

# -------- path generators --------

def make_path_linear(d: int, N: int) -> np.ndarray:
    ts = np.linspace(0.0, 1.0, N)
    path = np.empty((N, d), dtype=float)
    path[:, 0] = ts
    if d > 1:
        path[:, 1:] = 2.0 * ts[:, None]
    return path

def make_path_sin(d: int, N: int) -> np.ndarray:
    ts = np.linspace(0.0, 1.0, N)
    omega = 2.0 * math.pi
    ks = np.arange(1, d + 1, dtype=float)
    path = np.sin(omega * ts[:, None] * ks[None, :])
    return path

def make_path(d: int, N: int, kind: str) -> np.ndarray:
    if kind == "linear":
        return make_path_linear(d, N)
    elif kind == "sin":
        return make_path_sin(d, N)
    else:
        raise ValueError(f"Unknown path_kind: {kind}")

# -------- benchmarking helpers --------

def time_and_peak_memory(func, repeats: int = 5):
    best_time = float("inf")
    best_peak = 0

    for _ in range(repeats):
        tracemalloc.start()
        start = time.perf_counter()
        func()
        end = time.perf_counter()
        current, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()

        dur = end - start
        if dur < best_time:
            best_time = dur
            best_peak = peak

    return best_time, best_peak

# -------- benchmark implementations --------

def bench_iisignature(d: int, m: int, N: int, path_kind: str, operation: str, method: str, repeats: int):
    """Benchmark using iisignature library"""
    if not HAS_IISIG:
        return None
        
    # Force clear iisignature cache to prevent 'prepare' conflicts
    if hasattr(iisignature, "_basis_cache"):
        iisignature._basis_cache.clear()
        
    path = make_path(d, N, path_kind)
    
    if operation == "signature":
        arg = m
        func = iisignature.sig
    elif operation == "logsignature":
        if d < 2:
            return None
        try:
            arg = iisignature.prepare(d, m, method)
            func = lambda p, basis: iisignature.logsig(p, basis, method)
        except Exception as e:
            print(f"Error preparing iisignature for d={d}, m={m}, method={method}: {e}", file=sys.stderr)
            return None
    else:
        raise ValueError(f"Unknown operation: {operation}")

    try:
        # warmup
        _ = func(path, arg)

        def run_op():
            func(path, arg)

        t_sec, peak_bytes = time_and_peak_memory(run_op, repeats=repeats)
    except Exception as e:
        print(f"iisignature benchmark failed for {operation} d={d} m={m}: {e}", file=sys.stderr)
        return None

    t_ms = t_sec * 1000.0
    alloc_kib = peak_bytes / 1024.0

    return {
        "N": N,
        "d": d,
        "m": m,
        "path_kind": path_kind,
        "operation": operation,
        "library": "iisignature",
        "t_ms": t_ms,
        "alloc_KiB": alloc_kib,
    }

def bench_pysiglib(d: int, m: int, N: int, path_kind: str, operation: str, repeats: int):
    """Benchmark using pysiglib library (signature only - no logsig support)"""
    if not HAS_PYSIGLIB:
        return None
    
    # pysiglib only supports signature
    if operation != "signature":
        return None
    
    path = make_path(d, N, path_kind)
    
    try:
        # pysiglib API: pysiglib.signature(path, degree)
        func = lambda: pysiglib.signature(path, degree=m)
    except Exception as e:
        print(f"Error setting up pysiglib sig for d={d}, m={m}: {e}", file=sys.stderr)
        return None

    try:
        # warmup
        _ = func()

        t_sec, peak_bytes = time_and_peak_memory(func, repeats=repeats)
    except Exception as e:
        print(f"pysiglib benchmark failed for {operation} d={d} m={m}: {e}", file=sys.stderr)
        import traceback
        traceback.print_exc(file=sys.stderr)
        return None

    t_ms = t_sec * 1000.0
    alloc_kib = peak_bytes / 1024.0

    return {
        "N": N,
        "d": d,
        "m": m,
        "path_kind": path_kind,
        "operation": operation,
        "library": "pysiglib",
        "t_ms": t_ms,
        "alloc_KiB": alloc_kib,
    }

# -------- sweep + write grid to file --------

def run_bench():
    cfg = load_config()
    Ns = cfg["Ns"]
    Ds = cfg["Ds"]
    Ms = cfg["Ms"]
    path_kind = cfg["path_kind"]
    runs_dir = cfg["runs_dir"]
    repeats = cfg["repeats"]
    logsig_method = cfg["logsig_method"]
    operations = cfg["operations"]

    print("Running Python benchmark suite with config:")
    print(f"  path_kind     = {path_kind}")
    print(f"  Ns            = {Ns}")
    print(f"  Ds            = {Ds}")
    print(f"  Ms            = {Ms}")
    print(f"  operations    = {operations}")
    print(f"  runs_dir      = \"{runs_dir}\"")
    print(f"  repeats       = {repeats}")
    print(f"  logsig_method = \"{logsig_method}\" (iisignature only)")
    print(f"  iisignature   = {'available' if HAS_IISIG else 'NOT AVAILABLE'}")
    print(f"  pysiglib      = {'available' if HAS_PYSIGLIB else 'NOT AVAILABLE'}")

    script_dir = Path(__file__).resolve().parent

    # If BENCHMARK_RUN_DIR is set, we write everything there.
    env_run_dir = os.environ.get("BENCHMARK_RUN_DIR")
    if env_run_dir:
        runs_path = Path(env_run_dir)
    else:
        runs_path = script_dir / runs_dir

    runs_path.mkdir(parents=True, exist_ok=True)

    results = []

    for N in Ns:
        for d in Ds:
            for m in Ms:
                for op in operations:
                    # Benchmark iisignature
                    if HAS_IISIG:
                        res = bench_iisignature(d, m, N, path_kind, op,
                                                logsig_method, repeats=repeats)
                        if res is not None:
                            results.append(res)
                    
                    # Benchmark pysiglib (only for signature)
                    if HAS_PYSIGLIB and op == "signature":
                        res = bench_pysiglib(d, m, N, path_kind, op, repeats=repeats)
                        if res is not None:
                            results.append(res)

    ts = datetime.now().strftime("%Y%m%d_%H%M%S")
    csv_path = runs_path / f"run_python_{ts}.csv"

    fieldnames = ["N", "d", "m", "path_kind", "operation", "library", "t_ms", "alloc_KiB"]
    with csv_path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        for row in results:
            writer.writerow(row)

    print("=" * 60)
    print(f"Benchmark grid written to: {csv_path}")
    return csv_path

if __name__ == "__main__":
    run_bench()